wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+			newUniqueHostFunctions++
+		} else {
+			if _, ok := e.compiledWasmFunctionIndex[f]; ok {
+				continue
+			}
+			id := int64(len(e.compiledWasmFunctionIndex))
+			e.compiledWasmFunctionIndex[f] = id
+			newUniqueWasmFunctions++
+		}
+	}
+	e.hostFunctions = append(
+		e.hostFunctions,
+		make([]func(ctx *wasm.HostFunctionCallContext), newUniqueHostFunctions)...,
+	)
+	e.compiledWasmFunctions = append(
+		e.compiledWasmFunctions,
+		make([]*compiledWasmFunction, newUniqueWasmFunctions)...,
+	)
+	return nil
+}
+
+func (e *engine) Compile(f *wasm.FunctionInstance) error {
+	if f.HostFunction != nil {
+		id := e.hostFunctionIndex[f]
+		if e.hostFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		hf := func(ctx *wasm.HostFunctionCallContext) {
+			tp := f.HostFunction.Type()
+			in := make([]reflect.Value, tp.NumIn())
+			for i := len(in) - 1; i >= 1; i-- {
+				val := reflect.New(tp.In(i)).Elem()
+				raw := e.pop()
+				kind := tp.In(i).Kind()
+				switch kind {
+				case reflect.Float64, reflect.Float32:
+					val.SetFloat(math.Float64frombits(raw))
+				case reflect.Uint32, reflect.Uint64:
+					val.SetUint(raw)
+				case reflect.Int32, reflect.Int64:
+					val.SetInt(int64(raw))
+				}
+				in[i] = val
+			}
+			val := reflect.New(tp.In(0)).Elem()
+			val.Set(reflect.ValueOf(ctx))
+			in[0] = val
+			for _, ret := range f.HostFunction.Call(in) {
+				switch ret.Kind() {
+				case reflect.Float64, reflect.Float32:
+					e.push(math.Float64bits(ret.Float()))
+				case reflect.Uint32, reflect.Uint64:
+					e.push(ret.Uint())
+				case reflect.Int32, reflect.Int64:
+					e.push(uint64(ret.Int()))
+				default:
+					panic("invalid return type")
+				}
+			}
+		}
+		e.hostFunctions[id] = hf
+	} else {
+		id := e.compiledWasmFunctionIndex[f]
+		if e.compiledWasmFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		cf, err := e.compileWasmFunction(f)
+		if err != nil {
+			return fmt.Errorf("failed to compile Wasm function: %w", err)
+		}
+		e.compiledWasmFunctions[id] = cf
+	}
+	return nil
+}
+
+func NewEngine() wasm.Engine {
+	return newEngine()
+}
+
+const initialStackSize = 100
+
+func newEngine() *engine {
+	e := &engine{
+		stack:                     make([]uint64, initialStackSize),
+		compiledWasmFunctionIndex: make(map[*wasm.FunctionInstance]int64),
+		hostFunctionIndex:         make(map[*wasm.FunctionInstance]int64),
+	}
+	return e
+}
+
+func (e *engine) pop() (ret uint64) {
+	ret = e.stack[e.currentBaseStackPointer+e.currentStackPointer-1]
+	e.currentStackPointer--
+	return
+}
+
+func (e *engine) push(v uint64) {
+	e.stack[e.currentBaseStackPointer+e.currentStackPointer] = v
+	e.currentStackPointer++
+}
+
+type jitStatusCodes uint32
+
+const (
+	jitStatusReturned jitStatusCodes = iota
+	jitStatusCallWasmFunction
+	jitStatusCallBuiltInFunction
+	jitStatusCallHostFunction
+	// TODO: trap, etc?
+)
+
+var (
+	engineStackSliceOffset              = int64(unsafe.Offsetof((&engine{}).stack))
```

Comment:
since these are top-level fields, maybe we should describe what they are.

Are these mutable? if multiple engines are running, do these clash? or are multiple engines impossible?

Reply:
https://github.com/tetratelabs/wazero/pull/60/commits/a1b00ed9587c44b71bbf04e3d20d18f852b51499 done